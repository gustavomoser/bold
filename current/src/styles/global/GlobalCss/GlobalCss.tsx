import { Global as EmotionGlobal, Interpolation } from '@emotion/core'
import React from 'react'

import { useTheme } from '../../hooks'
import { Theme } from '../../theme/createTheme'

export const GlobalCss = EmotionGlobal

export const DefaultGlobalCss = () => {
  const theme = useTheme()

  return <GlobalCss styles={createGlobalStyles(theme)} />
}

export const createGlobalStyles = (theme: Theme): Interpolation => ({
  html: {
    fontSize: theme.typography.sizes.html,
    boxSizing: 'border-box',
  },
  '*, *:before, *:after': {
    boxSizing: 'inherit',
  },
  body: {
    margin: 0,
    color: theme.pallete.text.main,
    fontFamily: theme.typography.fontFamily,
    fontSize: theme.typography.sizes.text,
    overflowY: 'scroll',
  },
  hr: {
    backgroundColor: theme.pallete.divider,
    border: 'none',
    height: 1,
    margin: '1rem 0',
  },
  h1: { margin: 0, fontSize: '1.5rem' },
  h2: { margin: 0, fontSize: '1.25rem' },
  h3: { margin: 0, fontSize: '1rem' },
  h4: { margin: 0, fontSize: '0.875rem' },
  h5: { margin: 0, fontSize: '0.8125rem' },
  h6: { margin: 0, fontSize: '0.75rem' },
  p: {
    margin: '0',
    lineHeight: 1.5,
  },
})
