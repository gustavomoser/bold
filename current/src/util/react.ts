import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
