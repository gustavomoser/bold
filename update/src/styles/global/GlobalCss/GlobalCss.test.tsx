import { sheet } from '@emotion/css'
import { render } from '@testing-library/react'

import React from 'react'
import { DefaultGlobalCss, GlobalCss } from './GlobalCss'

describe('GlobalCss', () => {
  it('should inject global CSS styles', () => {
    render(<GlobalCss styles={{ body: { color: 'red' } }} />)
    expect(sheet.container).toMatchSnapshot()
  })
})

describe('DefaultGlobalCss', () => {
  it('should inject default CSS styles from theme', () => {
    render(<DefaultGlobalCss />)
    expect(sheet.container).toMatchSnapshot()
  })
})
