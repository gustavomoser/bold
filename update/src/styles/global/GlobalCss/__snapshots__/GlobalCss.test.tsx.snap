// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultGlobalCss should inject default CSS styles from theme 1`] = `
<head>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    html{font-size:16px;box-sizing:border-box;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    *,*:before,*:after{box-sizing:inherit;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    body{margin:0;color:#24252E;font-family:"IBM Plex Sans",sans-serif;font-size:0.8125rem;overflow-y:scroll;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    hr{background-color:#D3D4DD;border:none;height:1px;margin:1rem 0;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    h1{margin:0;font-size:1.5rem;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    h2{margin:0;font-size:1.25rem;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    h3{margin:0;font-size:1rem;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    h4{margin:0;font-size:0.875rem;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    h5{margin:0;font-size:0.8125rem;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    h6{margin:0;font-size:0.75rem;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    p{margin:0;line-height:1.5;}
  </style>
</head>
`;

exports[`GlobalCss should inject global CSS styles 1`] = `
<head>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    body{color:red;}
  </style>
</head>
`;
